File: CountAndSay.py
class Solution:
    # @return a string
    def countAndSay(self, n):
        if n == 1:
            return "1"
        last_str = self.countAndSay(n - 1)
        result = 0
        count = 0
        index = 0
        while index < len(last_str):
            if index == len(last_str) - 1:
                count += 1
                result *= 10
                result += count
                result *= 10
                result += int(last_str[index])
                break
            if last_str[index] == last_str[index + 1]:
                count += 1
                index += 1
            else:
                count += 1
                result *= 10
                result += count
                count = 0
                result *= 10
                result += int(last_str[index])
                index += 1
        return str(result)
                

File: CountAndSay.cc
#include <iostream>
#include <string>

using namespace std;

class Solution {
    public:
        string countAndSay(int n) {
            string result;
            int count = 0;
            while(count++ < n) {
                if (result.empty()) {
                    result = "1";
                    cout << "current iteration: " << result << endl;
                }
                else {
                    char current = result[0];
                    int char_counter = 1;
                    int index = 1;
                    string temp;
                    while (index < result.size()) {
                        if (result[index] == current) {
                            char_counter++;
                            index++;
                        } else {
                            temp += to_string(char_counter) + current;
                            current = result[index];
                            char_counter = 1;
                            index++;
                        }
                    }
                    temp += to_string(char_counter) + current;
                    result = temp;
                    cout << "current iteration: " << result << endl;
                }
            }
            return result;
        }
};

int main()
{
    Solution sol;
    cout << sol.countAndSay(4) << endl;
}
